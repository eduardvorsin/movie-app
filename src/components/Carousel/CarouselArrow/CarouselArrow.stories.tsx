import type { Meta, StoryObj } from '@storybook/react';
import CarouselArrow, { Props } from './CarouselArrow';
import { useTranslation } from 'react-i18next';

const meta: Meta<typeof CarouselArrow> = {
	title: 'UI/Carousel/CarouselArrow',
	tags: ['autodocs'],
	argTypes: {
		className: {
			description: 'Additional classes for the component'
		},
		testId: {
			description: 'Id for testing the component',
		},
		direction: {
			description: 'Allows you to set the direction of the arrow',
		},
		isDisabled: {
			description: 'Allows you to make the component inactive for interaction',
		},
		onClick: {
			control: false,
			description: 'Allows you to send a callback that will be called when clicked',
		},
		dictionary: {
			control: false,
			description: 'An object with keys for localization of the component',
		},
	},
	component: CarouselArrow,
	parameters: {
		docs: {
			description: {
				component: 'The CarouselArrow component is a control that provides navigation in the carousel component. This component provides arrows to move between the elements of the carousel, improving interactivity and user-friendliness.',
			},
		},
	},
}

export default meta;
type Story = StoryObj<typeof CarouselArrow>;

const CarouselArrowWithHooks = (props: Omit<Props, 'dictionary'>) => {
	const { t } = useTranslation('common');
	const dictionary = {
		direction: t('carouselArrow.direction', { context: props.direction })
	};

	return (
		<CarouselArrow
			{...props}
			dictionary={dictionary}
		/>
	);
}

export const Default: Story = {
	parameters: {
		docs: {
			description: {
				story: 'The default state of the component',
			},
		},
	},
	render: (args) => (
		<div className='flex gap-3 max-w-[150px] relative'>
			<CarouselArrowWithHooks {...args} direction='left' />
			<CarouselArrowWithHooks {...args} direction='right' />
		</div>
	),
};

export const Disabled: Story = {
	args: {
		direction: 'left',
		isDisabled: true,
	},
	parameters: {
		docs: {
			description: {
				story: 'Makes the component inactive for interaction',
			},
		},
	},
	render: (args) => (<CarouselArrowWithHooks {...args} />),
};

export const EventCallbacks: Story = {
	argTypes: {
		onClick: {
			action: 'Clicked',
		},
	},
	args: {
		direction: 'left',
	},
	parameters: {
		docs: {
			description: {
				story: 'Callback is a function that is called when clicking on a component',
			},
		},
	},
	render: (args) => (<CarouselArrowWithHooks {...args} />),
};