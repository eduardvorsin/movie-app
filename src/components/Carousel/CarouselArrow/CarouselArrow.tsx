import { MouseEventHandler } from 'react';
import { twMerge } from '../../../../tailwind.config';
import { GeneralProps } from '@/types/shared';

type Props = {
	direction: 'left' | 'right',
	isDisabled?: boolean,
	onClick: MouseEventHandler<HTMLButtonElement>,
} & GeneralProps;

export default function CarouselArrow({
	className,
	testId,
	isDisabled,
	direction,
	onClick,
	...props
}: Props) {
	const classes = twMerge(
		'p-2 md:p-1 absolute z-100 -translate-y-1/2 bg-neutral-300/80 enabled:hover:bg-neutral-300/[0.85] enabled:active:bg-neutral-300/90 dark:bg-dark-neutral-300/80 dark:enabled:hover:bg-dark-neutral-300/[0.85] dark:enabled:active:bg-dark-neutral-300/90 text-blue-700 enabled:hover:text-blue-800 enabled:active:text-blue-900 dark:text-blue-300 dark:enabled:hover:text-blue-400 dark:enabled:active:text-blue-500 transition-colors transition-opacity duration-150 disabled:opacity-disabled disabled:cursor-not-allowed',
		direction === 'left' ? 'left-[0.625rem]' : 'right-[0.625rem]',
		className
	);

	const iconFilename = direction === 'left' ? 'back-arrow.svg#back-arrow' : 'forward-arrow.svg#forward-arrow';

	return (
		<button
			className={classes}
			onClick={onClick}
			disabled={isDisabled}
			data-testid={testId}
			{...props}
		>
			<svg className='w-6 h-6 md:w-10 md:h-10' viewBox='0 0 32 32'>
				<use href={`/assets/icons/${iconFilename}`}></use>
			</svg>
			<span className='sr-only'>{direction.toUpperCase()} carousel control</span>
		</button>
	);
};