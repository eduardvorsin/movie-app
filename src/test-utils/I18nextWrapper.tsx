import React, { ReactNode } from 'react';
import i18next from '@/i18n/client';
import { I18nextProvider } from 'react-i18next';

type Props = {
	children: ReactNode,
}

const I18nextWrapper = ({
	children,
}: Props) => {
	return (
		<I18nextProvider i18n={i18next}>
			{children}
		</I18nextProvider>
	);
};

export default I18nextWrapper;