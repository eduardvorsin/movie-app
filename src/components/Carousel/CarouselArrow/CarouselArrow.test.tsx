import { render, screen } from "@testing-library/react";
import CarouselArrow, { Props } from "./CarouselArrow";
import userEvent from "@testing-library/user-event";
import i18next from "@/i18n/client";
import I18nextWrapper from "@/test-utils/I18nextWrapper";

const directions: NonNullable<Props['direction']>[] = ['left', 'right'];
const dictionary = { direction: 'Previous slide' };

describe('CarouselArrow tests', () => {
	it('is rendered correctly', () => {

		const mockFn = jest.fn();
		render(
			<CarouselArrow
				direction='left'
				dictionary={dictionary}
				onClick={mockFn}
				testId='test-carousel-arrow'
			/>
		);

		expect(screen.getByTestId<HTMLButtonElement>('test-carousel-arrow')).toBeInTheDocument();
	});

	it('when clicking on a component, the mock function is called', async () => {
		const user = userEvent.setup();
		const mockFn = jest.fn();
		render(
			<CarouselArrow
				direction='left'
				dictionary={dictionary}
				onClick={mockFn}
				testId='test-carousel-arrow'
			/>
		);

		await user.click(screen.getByRole<HTMLButtonElement>('button'));

		expect(mockFn).toHaveBeenCalledTimes(1);
	});

	it('when isDisabled is set to true, the component becomes inactive', async () => {
		const user = userEvent.setup();
		const mockFn = jest.fn();
		render(
			<CarouselArrow
				isDisabled
				direction='left'
				dictionary={dictionary}
				onClick={mockFn}
				testId='test-carousel-arrow'
			/>
		);

		await user.click(screen.getByRole<HTMLButtonElement>('button'));

		expect(mockFn).toHaveBeenCalledTimes(0);
	});

	it('is a basic snapshot', () => {
		const mockFn = jest.fn();
		render(
			<CarouselArrow
				direction='left'
				dictionary={dictionary}
				onClick={mockFn}
				testId='test-carousel-arrow'
			/>
		);

		expect(screen.getByTestId<HTMLButtonElement>('test-carousel-arrow')).toMatchSnapshot();
	});

	it('is a snapshot with isDisabled set to true', () => {
		const mockFn = jest.fn();
		render(
			<CarouselArrow
				isDisabled
				direction='left'
				dictionary={dictionary}
				onClick={mockFn}
				testId='test-carousel-arrow'
			/>
		);

		expect(screen.getByTestId<HTMLButtonElement>('test-carousel-arrow')).toMatchSnapshot();
	});

	it.each(directions)('a snapshot with direction equal to %s', (direction) => {
		const dictionary = { direction: direction === 'left' ? 'Previous slide' : 'Next slide' };
		const mockFn = jest.fn();
		render(
			<CarouselArrow
				isDisabled
				direction={direction}
				dictionary={dictionary}
				onClick={mockFn}
				testId='test-carousel-arrow'
			/>
		);

		expect(screen.getByTestId<HTMLButtonElement>('test-carousel-arrow')).toMatchSnapshot();
	});
});

describe('CarouselArrow integration tests', () => {
	it('localization into English works correctly', async () => {
		await i18next.changeLanguage('en');
		const translation = new RegExp(i18next.t('carouselArrow.direction', { context: 'left' }));

		const dictionary = {
			direction: i18next.t('carouselArrow.direction', { context: 'left' })
		};
		const mockFn = jest.fn();
		render(
			<CarouselArrow
				direction='left'
				dictionary={dictionary}
				onClick={mockFn}
				testId='test-carousel-arrow'
			/>,
			{ wrapper: I18nextWrapper }
		);

		expect(screen.getByText<HTMLSpanElement>(translation)).toBeInTheDocument();
	});

	it('localization into Russian works correctly', async () => {
		await i18next.changeLanguage('ru');
		const translation = new RegExp(i18next.t('carouselArrow.direction', { context: 'left' }));

		const dictionary = {
			direction: i18next.t('carouselArrow.direction', { context: 'left' })
		};
		const mockFn = jest.fn();
		render(
			<CarouselArrow
				direction='left'
				dictionary={dictionary}
				onClick={mockFn}
				testId='test-carousel-arrow'
			/>,
			{ wrapper: I18nextWrapper }
		);

		expect(screen.getByText<HTMLSpanElement>(translation)).toBeInTheDocument();
	});

	it('is a snapshot with English localization', async () => {
		await i18next.changeLanguage('en');

		const dictionary = {
			direction: i18next.t('carouselArrow.direction', { context: 'left' })
		};
		const mockFn = jest.fn();
		render(
			<CarouselArrow
				direction='left'
				dictionary={dictionary}
				onClick={mockFn}
				testId='test-carousel-arrow'
			/>,
			{ wrapper: I18nextWrapper }
		);

		expect(screen.getByTestId<HTMLButtonElement>('test-carousel-arrow')).toMatchSnapshot();
	});

	it('is a snapshot with Russian localization', async () => {
		await i18next.changeLanguage('ru');

		const dictionary = {
			direction: i18next.t('carouselArrow.direction', { context: 'left' })
		};
		const mockFn = jest.fn();
		render(
			<CarouselArrow
				direction='left'
				dictionary={dictionary}
				onClick={mockFn}
				testId='test-carousel-arrow'
			/>,
			{ wrapper: I18nextWrapper }
		);

		expect(screen.getByTestId<HTMLButtonElement>('test-carousel-arrow')).toMatchSnapshot();
	});
});