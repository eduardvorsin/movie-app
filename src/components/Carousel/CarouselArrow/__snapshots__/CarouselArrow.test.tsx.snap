// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CarouselArrow integration tests is a snapshot with English localization 1`] = `
<button
  class="p-2 md:p-1 absolute z-100 -translate-y-1/2 bg-neutral-300/80 enabled:hover:bg-neutral-300/[0.85] enabled:active:bg-neutral-300/90 dark:bg-dark-neutral-300/80 dark:enabled:hover:bg-dark-neutral-300/[0.85] dark:enabled:active:bg-dark-neutral-300/90 text-blue-700 enabled:hover:text-blue-800 enabled:active:text-blue-900 dark:text-blue-300 dark:enabled:hover:text-blue-400 dark:enabled:active:text-blue-500 transition-[color,background-color,opacity] duration-150 disabled:opacity-disabled disabled:cursor-not-allowed left-[0.625rem] "
  data-testid="test-carousel-arrow"
>
  <svg
    class="w-6 h-6 md:w-10 md:h-10"
    viewBox="0 0u 32 32"
  >
    <use
      href="/assets/icons/back-arrow.svg#back-arrow"
    />
  </svg>
  <span
    class="sr-only"
  >
    Previous slide
  </span>
</button>
`;

exports[`CarouselArrow integration tests is a snapshot with Russian localization 1`] = `
<button
  class="p-2 md:p-1 absolute z-100 -translate-y-1/2 bg-neutral-300/80 enabled:hover:bg-neutral-300/[0.85] enabled:active:bg-neutral-300/90 dark:bg-dark-neutral-300/80 dark:enabled:hover:bg-dark-neutral-300/[0.85] dark:enabled:active:bg-dark-neutral-300/90 text-blue-700 enabled:hover:text-blue-800 enabled:active:text-blue-900 dark:text-blue-300 dark:enabled:hover:text-blue-400 dark:enabled:active:text-blue-500 transition-[color,background-color,opacity] duration-150 disabled:opacity-disabled disabled:cursor-not-allowed left-[0.625rem] "
  data-testid="test-carousel-arrow"
>
  <svg
    class="w-6 h-6 md:w-10 md:h-10"
    viewBox="0 0u 32 32"
  >
    <use
      href="/assets/icons/back-arrow.svg#back-arrow"
    />
  </svg>
  <span
    class="sr-only"
  >
    Предыдущий слайд
  </span>
</button>
`;

exports[`CarouselArrow tests a snapshot with direction equal to left 1`] = `
<button
  class="p-2 md:p-1 absolute z-100 -translate-y-1/2 bg-neutral-300/80 enabled:hover:bg-neutral-300/[0.85] enabled:active:bg-neutral-300/90 dark:bg-dark-neutral-300/80 dark:enabled:hover:bg-dark-neutral-300/[0.85] dark:enabled:active:bg-dark-neutral-300/90 text-blue-700 enabled:hover:text-blue-800 enabled:active:text-blue-900 dark:text-blue-300 dark:enabled:hover:text-blue-400 dark:enabled:active:text-blue-500 transition-[color,background-color,opacity] duration-150 disabled:opacity-disabled disabled:cursor-not-allowed left-[0.625rem] "
  data-testid="test-carousel-arrow"
  disabled=""
>
  <svg
    class="w-6 h-6 md:w-10 md:h-10"
    viewBox="0 0u 32 32"
  >
    <use
      href="/assets/icons/back-arrow.svg#back-arrow"
    />
  </svg>
  <span
    class="sr-only"
  >
    Previous slide
  </span>
</button>
`;

exports[`CarouselArrow tests a snapshot with direction equal to right 1`] = `
<button
  class="p-2 md:p-1 absolute z-100 -translate-y-1/2 bg-neutral-300/80 enabled:hover:bg-neutral-300/[0.85] enabled:active:bg-neutral-300/90 dark:bg-dark-neutral-300/80 dark:enabled:hover:bg-dark-neutral-300/[0.85] dark:enabled:active:bg-dark-neutral-300/90 text-blue-700 enabled:hover:text-blue-800 enabled:active:text-blue-900 dark:text-blue-300 dark:enabled:hover:text-blue-400 dark:enabled:active:text-blue-500 transition-[color,background-color,opacity] duration-150 disabled:opacity-disabled disabled:cursor-not-allowed right-[0.625rem] "
  data-testid="test-carousel-arrow"
  disabled=""
>
  <svg
    class="w-6 h-6 md:w-10 md:h-10"
    viewBox="0 0u 32 32"
  >
    <use
      href="/assets/icons/forward-arrow.svg#forward-arrow"
    />
  </svg>
  <span
    class="sr-only"
  >
    Next slide
  </span>
</button>
`;

exports[`CarouselArrow tests is a basic snapshot 1`] = `
<button
  class="p-2 md:p-1 absolute z-100 -translate-y-1/2 bg-neutral-300/80 enabled:hover:bg-neutral-300/[0.85] enabled:active:bg-neutral-300/90 dark:bg-dark-neutral-300/80 dark:enabled:hover:bg-dark-neutral-300/[0.85] dark:enabled:active:bg-dark-neutral-300/90 text-blue-700 enabled:hover:text-blue-800 enabled:active:text-blue-900 dark:text-blue-300 dark:enabled:hover:text-blue-400 dark:enabled:active:text-blue-500 transition-[color,background-color,opacity] duration-150 disabled:opacity-disabled disabled:cursor-not-allowed left-[0.625rem] "
  data-testid="test-carousel-arrow"
>
  <svg
    class="w-6 h-6 md:w-10 md:h-10"
    viewBox="0 0u 32 32"
  >
    <use
      href="/assets/icons/back-arrow.svg#back-arrow"
    />
  </svg>
  <span
    class="sr-only"
  >
    Previous slide
  </span>
</button>
`;

exports[`CarouselArrow tests is a snapshot with isDisabled set to true 1`] = `
<button
  class="p-2 md:p-1 absolute z-100 -translate-y-1/2 bg-neutral-300/80 enabled:hover:bg-neutral-300/[0.85] enabled:active:bg-neutral-300/90 dark:bg-dark-neutral-300/80 dark:enabled:hover:bg-dark-neutral-300/[0.85] dark:enabled:active:bg-dark-neutral-300/90 text-blue-700 enabled:hover:text-blue-800 enabled:active:text-blue-900 dark:text-blue-300 dark:enabled:hover:text-blue-400 dark:enabled:active:text-blue-500 transition-[color,background-color,opacity] duration-150 disabled:opacity-disabled disabled:cursor-not-allowed left-[0.625rem] "
  data-testid="test-carousel-arrow"
  disabled=""
>
  <svg
    class="w-6 h-6 md:w-10 md:h-10"
    viewBox="0 0u 32 32"
  >
    <use
      href="/assets/icons/back-arrow.svg#back-arrow"
    />
  </svg>
  <span
    class="sr-only"
  >
    Previous slide
  </span>
</button>
`;
